### PR TITLE
sql: fix NULL handling for IN and IS (NOT) DISTINCT FROM for (non-nested) tuples

### PR DESCRIPTION
**Pull request by RaduBerinde**
#### sql: fix NULL handling in tuple IN expressions

This change fixes handling of NULLs with IN expressions. The following
examples illustrate the interesting cases that are fixed:
```
  NULL IN <empty-tuple>       = false
  (1, NULL) IN <empty-tuple>  = false
  (1, NULL) IN ((1, NULL))    = NULL
  (1, NULL) IN ((1, 1))       = NULL
  (1, NULL) IN ((2, NULL))    = false
```

To implement this correctly, we now require that only tuples that
don't contain NULLs can be marked as sorted. This is because we can't
use binary search to, for example "match" `(1, NULL)` with `(1, 1)`
(necessary to know if the result is false or NULL).

Note that equality is broken for deeper tuples (#22717), and the new
IN code relies on the equality code so it is broken for deep tuples as
well.

Updates #22670.

Release note: None

#### sql: refactor handing of comparison ops that tolerate NULLs

We have a bunch of special cases around comparison ops that tolerate
NULLs, like `IN` and `IS (NOT) DISTINCT FROM`. This refactoring cleans
this up as follows:
 - create `CmpOps` for `IsNotDistinctFrom`
 - add a `CmpOp.nullableArgs` field to indicate if the operation
   tolerates NULLs, replacing the special-casing of the operator.
 - we refactor the helper functions in the comparison type check code
   to return an `alwaysNull` flag, instead of relying on "matching"
   checks inside and outside the functions.

Release note: None

#### sql: fix IS (NOT) DISTINCT FROM for tuples

Fixing handling for cases like `(1, NULL) IS NOT DISTINCT FROM (1,
NULL)`, which currently returns NULL.

This requires removing the normalization rules for IS (NOT) DISTINCT
FROM. These rules are not correct when the arguments are tuples. They
were introduced to tickle the old index constraints code into
generating spans; the new code handles IS (NOT) DISTINCT directly so
these are no longer needed (adding some logic tests to illustrate
this).

Updates #22670.

Release note: None

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Note that fixing index constraint generation for tuples that contain NULLs will be addressed in a separate PR.

**Comment by RaduBerinde** [comment]
Note: SimplifyExpr is not used for index constraints anymore. There is only one (dubious) usage in the partitioning code. I am filing a separate bug to investigate that.

**Comment by knz** [comment]
Thank you for the fast turnaround.

LGTM modulo:

1) Before this can merge each of the 3  commits needs a "bug fix" release note.

2) kind request for 3 additional comments in the code

---

Reviewed 8 of 8 files at r1, 9 of 9 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sem/tree/eval.go, line 1939 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-L5Lj8j4OeLdqGthRiUn:-L5Lj8j4OeLdqGthRiUo:bwy1e2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffacee9d2f13b445f897129d08c759f8a055ddc5/pkg/sql/sem/tree/eval.go#L1939)):*
> ```Go
> 			argTuple, argIsTuple := arg.(*DTuple)
> 			if vtuple.Sorted() && !(argIsTuple && argTuple.ContainsNull()) {
> 				_, result := vtuple.SearchSorted(ctx, arg)
> ```

Provide an example in a comment that shows what gets into this if branch.

---

*[pkg/sql/sem/tree/eval.go, line 1945 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-L5LjE_6ofyXs_7hdPmI:-L5LjE_6ofyXs_7hdPmJ:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffacee9d2f13b445f897129d08c759f8a055ddc5/pkg/sql/sem/tree/eval.go#L1945)):*
> ```Go
> 			sawNull := false
> 			if !argIsTuple {
> 				for _, val := range vtuple.D {
> ```

ditto

---

*[pkg/sql/sem/tree/eval.go, line 1953 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-L5LjFPFi6C4fKg791eC:-L5LjFPGrlSaRgNMJbgW:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffacee9d2f13b445f897129d08c759f8a055ddc5/pkg/sql/sem/tree/eval.go#L1953)):*
> ```Go
> 				}
> 			} else {
> 				for _, val := range vtuple.D {
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-:-L5Lid5eu82aad8FW40i:blthtl3)*

**Comment by RaduBerinde** [comment]
TFTR! Added release notes (thanks for cathcing that!).

---

Review status: 4 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/sem/tree/eval.go, line 1939 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-L5Lj8j4OeLdqGthRiUn:-L5LnSvRLhSqGint2ZZE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffacee9d2f13b445f897129d08c759f8a055ddc5/pkg/sql/sem/tree/eval.go#L1939)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Provide an example in a comment that shows what gets into this if branch.
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 1945 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-L5LjE_6ofyXs_7hdPmI:-L5LnTAYlRa_OkmWB20g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffacee9d2f13b445f897129d08c759f8a055ddc5/pkg/sql/sem/tree/eval.go#L1945)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/sem/tree/eval.go, line 1953 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-L5LjFPFi6C4fKg791eC:-L5LnTneJ1iZcMl2Xsnb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ffacee9d2f13b445f897129d08c759f8a055ddc5/pkg/sql/sem/tree/eval.go#L1953)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-:-L5LnQ8D1VCrCVc_G0g2:b-mfg9we)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 5 of 9 files at r5, 6 of 6 files at r6.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-:-L5Lu_EWUPc3YM9YQjmO:bnfp4nl)*

**Comment by jordanlewis** [comment]
:lgtm:

---

Reviewed 4 of 8 files at r1, 11 of 11 files at r4, 9 of 9 files at r5, 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/sem/tree/datum.go, line 2517 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-L5OwXSKwiiCNDGO97L6:-L5OwXSLq4n80HGagFSH:b-vqkp3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2db0762c85834427ce146b8d8f6aa17f78ebf431/pkg/sql/sem/tree/datum.go#L2517)):*
> ```Go
> 	d.AssertSorted()
> 	if target == DNull {
> 		panic("NULL target")
> ```

maybe include the left side in this panic message as well? fingers crossed we'll never hit this, but better safe than sorry

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22718#-:-L5Oyqw3sY6bcN8JojWJ:bnfp4nl)*

File: pkg/sql/sem/tree/datum.go [hunk 1]
```diff
@@ -2542,6 +2562,23 @@ func (d *DTuple) Size() uintptr {
 	return sz
 }
 
+// ContainsNull returns true if the tuple contains NULL, possibly nested inside
+// other tuples. For example, all the following tuples contain NULL:
+//
```
Comment (nvanbenschoten) [inline pkg/sql/sem/tree/datum.go]: What following tuples?

File: pkg/sql/sem/tree/eval.go [hunk 2]
```diff
@@ -1922,38 +1922,56 @@ func makeEvalTupleIn(typ types.T) CmpOp {
...
+						sawNull = true
+					} else if val.Compare(ctx, arg) == 0 {
+						return DBoolTrue, nil
+					}
+				}
+			} else {
+				// The left-hand side is a tuple, e.g. `(1, 2) IN ((1, 2), (3, 4))`.
+				for _, val := range vtuple.D {
+					var res Datum
+					if argIsTuple {
```
Comment (nvanbenschoten) [inline pkg/sql/sem/tree/eval.go]: Isn't this always true?

**Review by nvb — COMMENTED** [review]
LGTM. Sorry for the late review. Your call as to whether these two comments are worth a new PR.

**Comment by RaduBerinde** [comment]
Thanks, opened https://github.com/cockroachdb/cockroach/pull/22744

**Pull request merged**